src/common/base/abstractAsyncOperation.ts

```diff
@@ -0,0 +1,109 @@
+import { Duration } from '@salesforce/kit';
+import { DeployPipelineCache } from '../deployPipelineCache';
+import DoceMonitor from '../../streamer/doceMonitor';
+import { AsyncOperationResult, AsyncOperationResultJson, AsyncOperationStatus } from '../types';
+import { fetchAsyncOperationResult, getAsyncOperationStreamer } from '../../common';
+import { PromoteOutputService } from '../outputService';
+
+Messages.importMessagesDirectory(__dirname);
+const messages = Messages.loadMessages('@salesforce/plugin-devops-center', 'commonErrors');
+
+export abstract class AsyncCommand extends SfCommand<AsyncOperationResultJson> {
+  protected targetOrg: Org;
+  private asyncOperationId: string;
+  private outputService: PromoteOutputService;
+
+  protected abstract baseCommand: string;
+
+  protected setAsyncOperationId(aorId: string): void {
+    this.asyncOperationId = aorId;
+  }
+
+  protected setOutputService(service: PromoteOutputService): void {
+    this.outputService = service;
+  }
+
+  /**
+   * Method to handle an in progress async operation. It stores the AOR in cache,
+   * then monitors the state of it and print messages if needed. Finally returns the
+   * final state of the operation.
+   */
+  protected async monitorOperation(isAsync: boolean, wait: Duration): Promise<AsyncOperationResultJson> {
+    await this.cacheOperation();
+    this.outputService.setAorId(this.asyncOperationId);
+    this.outputService.printAorId();
+    this.outputService.printOpSummary();
+
+    if (isAsync) {
+      return this.getOperationState();
+    }
+
+    await this.startMonitoring(wait);
+    if (this.outputService.getStatus() === AsyncOperationStatus.Completed) {
+      this.outputService.displayEndResults();
+    }
+    return this.getOperationState();
+  }
+
+  /**
+   * Default function for catching commands errors.
+   *
+   * @param error
+   * @returns
+   */
+  protected catch(error: Error | SfError): Promise<SfCommand.Error> {
+    if (error.name.includes('GenericTimeoutError')) {
+      const err = messages.createError('error.ClientTimeout', [this.config.bin, this.baseCommand]);
+      return super.catch({ ...error, name: err.name, message: err.message, code: err.code });
+    }
+    return super.catch(error);
+  }
+
+  /**
+   *
+   * Stores the AOR Id in cache.
+   */
+  protected async cacheOperation(): Promise<void> {
+    await DeployPipelineCache.set(this.asyncOperationId, {});
+  }
+
+  /**
+   *
+   * Executes a query to get the state of the AOR record.
+   */
+  private async getOperationState(): Promise<AsyncOperationResultJson> {
+    const asyncJob: AsyncOperationResult = await fetchAsyncOperationResult(
+      this.targetOrg.getConnection(),
+      this.asyncOperationId
+    );
+    return {
+      jobId: this.asyncOperationId,
+      status: asyncJob.sf_devops__Status__c as string,
+      message: asyncJob.sf_devops__Message__c as string,
+      errorDetails: asyncJob.sf_devops__Error_Details__c as string,
```

Comment:
If the errorDetails here is undefined or null you will get an error here I think.
There are some cases where Heroku throws an error and you don't get the Error_details.

Reply:
That would be similar to a successful promotion, where  _sf_devops__Error_Details__c_ is null right? I tested it and it doesn't throw.

Reply:
Yes, you will have and status = ERROR and no Error_Details at all.
I had that issue and it doesn’t print anything on the screen about the error because it was not handled on aorOutputService printAorStatus.
On my future PR I have a fix for that, we are going to print the error even if we don’t have error details.